File: LEET_1631.cpp
class Solution {
public:
    int R;
    int C;
    int xmoves[4] = {0,0,1,-1};
    int ymoves[4] = {1,-1,0,0};
    
    int minimumEffortPath(vector<vector<int>>& heights) {
        R = heights.size();
        C = heights[0].size();
        int left=0, right=0;
        
        // set right at the maximum
        for (auto& row: heights){
            for (int h: row){
                right = max(right, h);
            }
        }
        
        while (left <= right){
            int mid = left + (right-left)/2;
            if (check(mid, heights)){
                // since we're looking for the minimum
                right = mid-1;
            } else {
                left = mid+1;
            }
        }
        return left;
    }
    
    bool check(int effort, vector<vector<int>>& heights){
        vector<vector<bool>> seen(R, vector<bool>(C, false));
        stack<pair<int,int>> st;
        seen[0][0] = true;
        st.push(pair(0,0));
        
        while (!st.empty()){
            auto [cx, cy] = st.top();
            st.pop();
            
            if (cx==R-1 && cy==C-1) return true;
            
            
            for (int idx=0; idx<4; idx++) {
                int nx = cx + xmoves[idx];
                int ny = cy + ymoves[idx];
                if (0<=nx && nx<R && 0<=ny && ny<C && !seen[nx][ny]){
                    if (abs(heights[nx][ny]-heights[cx][cy]) <= effort){
                        st.push(pair(nx,ny));
                        seen[nx][ny] = true;
                    }
                }
            }
        }
        return false;
    }
};
//     // 1st try: Time limit exceeded!! 
//     int diff = INT_MAX; // should be smallest

//     void step(int cx, int cy, int max_effort, vector<vector<int>>& heights, vector<vector<int>>& visited) {
//         int X = heights.size(), Y = heights[0].size();
//         int xmoves[4] = {0, 0, -1, 1};
//         int ymoves[4] = {1, -1, 0, 0};

//         if (cx == X - 1 && cy == Y - 1) {
//             diff = min(max_effort, diff);
//             return;
//         }

//         for (int i = 0; i < 4; i++) {
//             int nx = cx + xmoves[i];
//             int ny = cy + ymoves[i];
//             if (0 <= nx && nx < X && 0 <= ny && ny < Y && visited[nx][ny] == 0) {
//                 visited[nx][ny] = 1;
//                 int effort = abs(heights[nx][ny] - heights[cx][cy]);
//                 int temp_max_effort = max(max_effort, effort);
//                 if (temp_max_effort > diff) continue;
//                 step(nx, ny, temp_max_effort, heights, visited);
//                 visited[nx][ny] = 0;
//             }
//         }
//     }

//     int minimumEffortPath(vector<vector<int>>& heights) {
//         int X = heights.size(), Y = heights[0].size();
//         vector<vector<int>> visited(X, vector<int>(Y, 0));

//         visited[0][0] = 1;
//         step(0, 0, 0, heights, visited);
//         return diff;
//     }    
    

File: LEET_1631.py
"""
1631. Path With Minimum Effort
"""


from typing import List


class Solution:
    def minimumEffortPath(self, heights: List[List[int]]) -> int:
        m, n = len(heights), len(heights[0])
        dy = [-1, 1, 0, 0]
        dx = [0, 0, -1, 1]

        def dfs(y, x, visited, threadshold):
            if y == m - 1 and x == n - 1:
                return True  # Reach destination

            visited[y][x] = 1
            for d in range(4):
                ny = y + dy[d]
                nx = x + dx[d]
                if not (0 <= ny < m and 0 <= nx < n and visited[ny][nx] == 0):
                    continue
                if abs(heights[ny][nx] - heights[y][x]) <= threadshold and dfs(ny, nx, visited, threadshold):
                    return True
            return False

        def canReachDestination(threadshold):
            visited = [[0] * n for _ in range(m)]
            return dfs(0, 0, visited, threadshold)

        left = 0
        ans = right = 10**6
        while left <= right:
            # 반올림과 버림의 차이?
            mid = left + (right - left) // 2
            if canReachDestination(mid):
                right = mid - 1  # Try to find better result on the left side
                ans = mid
            else:
                left = mid + 1
        return ans


if __name__ == "__main__":
    solution = Solution()
    heights = [
        [[1, 2, 2], [3, 8, 2], [5, 3, 5]],
        [[1, 2, 3], [3, 8, 4], [5, 3, 5]],
        [[1, 2, 1, 1, 1], [1, 2, 1, 2, 1], [1, 2, 1, 2, 1], [1, 2, 1, 2, 1], [1, 1, 1, 2, 1]],
    ]
    output = [2, 1, 0]

    for h in heights:
        print(solution.minimumEffortPath(h))
